docs/sovereign/standalone-evm.md

```diff
@@ -0,0 +1,172 @@
+# Standalone EVM
+
+## EVM as example
+In the early stages of the MultiversX VM development, there were already components built specifically for EVM compatibility. We are revisiting and reusing parts of that code. In **VM1.2**, for instance, there was a direct correspondence between EVM opcodes and the **BlockchainHook** interface, as well as a mechanism that wrapped MvX-style transaction data (**txData**) into EVM-specific `vmInput`.
+
+---
+
+## 1. VMExecutionHandlerInterface
+The MultiversX protocol defines a **VMExecutionHandlerInterface** with the following functions:
+
+```go
+// RunSmartContractCreate computes how a smart contract creation should be performed
+RunSmartContractCreate(input *ContractCreateInput) (*VMOutput, error)
+
+// RunSmartContractCall computes the result of a smart contract call and how the system must change after the execution
+RunSmartContractCall(input *ContractCallInput) (*VMOutput, error)
+```
+The **SCProcessor** from `mx-chain-go` prepares the input information for these functions. We aim to avoid modifying the **SCProcessor** itself; instead, all necessary abstractions will be implemented at the EVM level.
+
+---
+
+## 2. Input Preparation: EVMInputCreator
+
+When a contract creation request is made (via *ContractCreateInput), an EVMInputCreator component will:
+- Convert the `ContractCreateInput` into an EVMInput.
+- Invoke the actual EVM smart contract logic.
+
+The EVM itself is taken from the official Go implementation ([evm.go](https://github.com/ethereum/go-ethereum/blob/master/core/vm/evm.go) in go-ethereum).
+
+---
+
+## 3. Abstraction Layer: MultiversX & EVM Interfaces
+
+To allow the EVM to function within MultiversX, we introduce a layer that bridges EVM interfaces with MultiversX components. The core interface it uses is the `BlockchainHookInterface`, which grants access to critical blockchain data, state, and transaction information.
+### 3.1 Reading & Writing to Storage
+
+- **Reading Storage**: When an EVM opcode attempts to read data from the storage (e.g., `readStorageFromTrie(key)`), it should invoke `blockchainHook.ReadFromStorage(scAddress, key)`.
+        Internally, this call goes through the `storageContext` component, which manages reads from local cache if a key has already been accessed or modified during the current transaction.
+
+- **Writing to Storage**: When writing to storage, the EVM opcode should call `SetStorageToAddress(address, key)` in the `storageContext`.
+
+### 3.2 Finalizing State Changes
+
+After EVM execution finishes, we need to commit the resulting state changes to the blockchain. The EVM will use the `outputContext` component, which (together with the `storageContext`) tracks modified accounts and storages. It also creates the final `vmOutput`, which the `scProcessor` in `mx-chain-go` will then validate and apply to the blockchain (the trie) if everything is correct.
+
+---
```

Comment:
I think we didn't use the separators `---` before. Maybe remove them?

Reply:
:+1: